core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java

```diff
@@ -426,6 +429,31 @@ public ProxiedAddresses proxiedAddresses() {
         return proxiedAddresses;
     }
 
+    @Override
+    public CompletableFuture<Void> initiateConnectionShutdown() {
```

Comment:
Armeria has its own mechanism to manage connections with [KeepAliveHander]https://github.com/line/armeria/blob/32eabd2eb1f1be73f03a6425fa193631064cf2db/core/src/main/java/com/linecorp/armeria/internal/common/KeepAliveHandler.java]. 
I think we can leverage `KeepAliveHander.needToCloseConnection()` to send GOAWAY frame for HTTP/2 and "Connection: close" for HTTP/1 rather than directly access `Channel` API
https://github.com/line/armeria/blob/32eabd2eb1f1be73f03a6425fa193631064cf2db/core/src/main/java/com/linecorp/armeria/internal/common/KeepAliveHandler.java#L62-L65

However, `DefaultServiceRequestContext` can not have a `KeepAliveHandler` instance.
So we need to make `DefaultServiceRequestContext` access a `KeepAliveHander`:
You may want to see and refer to the following code points and steps. 
1. A `KeepAliveHander` is passed to `DecodedHttpRequest` when it is created.
  * For HTTP/2
  https://github.com/line/armeria/blob/89a1eb80a0e7b1e4d8bc505632e3b3b176891f52/core/src/main/java/com/linecorp/armeria/server/Http2RequestDecoder.java#L149-L158
  * For HTTP/1
  https://github.com/line/armeria/blob/e29b577e57dd02f4f24d76ac2baa39447bf13d30/core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java#L211-L220
2. The `KeepAliveHander` in `DecodedHttpRequest` can be injected to `DefaultServiceRequestContext` from:
https://github.com/line/armeria/blob/cc495bcae11a37dfc8a2d2e166dbf8c1f57fb552/core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java#L381-L385

Reply:
Thanks for the detailed pointers, @ikhoon, super helpful! I drafted some changes following your suggestions in a separate commit, pushed it for you to take a look and get feedback, will clean it up and update the tests once we settle on a final implementation. A couple of questions:

1. Is it okay to change constructor signature for DefaultDecodedHttpRequest and EmptyContentDecodedHttpRequest? They seem to be a part of public API.

2. It seems like `KeepAliveHander.needToCloseConnection()` is causing GOAWAY frame to be sent from the `ServerHttp2ObjectEncoder.doWriteHeaders` for HTTP/2, which means it's not going to be sent until handler starts sending response. This is not ideal - we might need to send GOAWAY frame immediately once the `initiateConnectionShutdown` is called, to make sure client receives it ASAP without waiting for the handler to start sending headers, which depending on implementation and task could be delayed significantly. HTTP/2 specification revolves around the idea of concurrent exchanges, server should be able to send GOAWAY frame for the connection without being blocked on processing of other streams. Any thoughts?

https://github.com/line/armeria/blob/68b9e243aff5aa67ef878076a13bbb5f435ca5ae/core/src/main/java/com/linecorp/armeria/server/ServerHttp2ObjectEncoder.java#L69-L74

3. Another thing to take into account - as mentioned in the "Potential follow-ups" section in the PR description, server can send 2 GOAWAY frames, 1st frame with stream ID 2^31-1 to signal shutdown intent to the client without rejecting new streams, and, after some grace period, 2nd frame with latest accepted/processed stream ID and start rejecting any incoming streams with higher stream IDs w/ RST_STREAM frame. What would be your recommendation on how to fit this into the `KeepAliveHandler` functionality?

I think, in general it makes sense to implement this using `KeepAliveHandler` - no need to detect which protocol is handled inside the `DefaultServiceRequestContext`, as a nice bonus it provides support for HTTP/1 out of the box. However, to address 2. and 3. for sending HTTP/2 frames concurrently with handle we might need to use Channel API directly from KeepAliveHandler and either:

a) override `destroy()` for `Http2ServerKeepAliveHandler` (currently marked as final in `AbstractKeepAliveHandler`) to send GOAWAY instantly for 2. and schedule follow up frame for 3. with some delay (also, make it configurable).

or 

b) provide a new method for doing this in `KeepAliveHandler` interface, but I'm not sure if that's a good idea since it's not that useful for all the class that implement this interface.

Let me know what you think and if you have any other ideas.

Reply:
> 1. Is it okay to change constructor signature for DefaultDecodedHttpRequest and EmptyContentDecodedHttpRequest? They seem to be a part of public API.

Yes, you can do it. Because they are package-private classes that do not `pubilc` keyword in the class definition.

> 2. This is not ideal - we might need to send GOAWAY frame immediately once the initiateConnectionShutdown is called, to make sure client receives it ASAP without waiting for the handler to start sending headers,

That makes sense. How about making `Http2ServerKeepAliveHandler.needToCloseConnection()` send a GOAWAY frame immediately by overriding super class, `isGoAwaySent` flag will prohibit encoder() from writing duplicate GOAWAY frame.

> 3. ... after some grace period, 2nd frame with latest accepted/processed stream ID and start rejecting any incoming streams with higher stream IDs w/ RST_STREAM frame.

IIUC, if a new stream ID greater than `lastKnownStreamId` used for sending GOAWAY frame is received, the new stream will be rejected by Netty HTTP/2 codec.
https://github.com/netty/netty/blob/7971a8ca49d97a9a09bb2f936453fd3ef3d8c184/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L634-L647

Reply:
Sending GOAWAY frame from `needToCloseConnection` doesn't address the issue because `ServerHttp2ObjectEncoder.doWriteHeaders` is called only when response headers are ready to be sent.

To avoid any confusion I just went ahead and added implementation of the grace period as described above, GOAWAY frames are written from inside KeepAliveHandler. 

For HTTP/2
- If grace period is positive - sends GOAWAY frame with stream ID 2^31-1, then after grace period sends GOAWAY with connection().remote().lastStreamCreated(). 
- If grace period is zero or negative - immediately sends GOAWAY with connection().remote().lastStreamCreated()

For HTTP1
- If grace period is positive - schedules KeepAliveHandler::destroy after grace period. 
- If grace period is zero or negative - immediately calls KeepAliveHandler::destroy.
- In both cases it results in "Connection: close" header added in `ServerHttp1ObjectEncoder.doWriteHeaders`

Please take a look, let me know what you think

Reply:
So far I see a couple of problems with adding KeepAliveHandler to the DefaultServiceRequestContext:

1. `DefaultServiceRequestContext` methods can be called concurrently from multiple threads - while KeepAliveHandler implementations don't seem to be thread-safe in general. If we're calling `KeepAliveHandler` methods from `DefaultServiceRequestContext` we'll need to add synchronization for all internal state for all implementations of `KeepAliveHandler`. I mentioned `isFinalGoAwaySent` in the comment above, also I'm calling `destroy` from inside `initiateConnectionShutdown`, so [isInitialized](https://github.com/line/armeria/blob/467da571eabee80b7f55ea8b6bf2ed839c91f508/core/src/main/java/com/linecorp/armeria/internal/common/AbstractKeepAliveHandler.java#L148) should be also synchronized to protect from races.

2. CI checks are unhappy:

```
* What went wrong:
Execution failed for task ':javadoc:checkJavadoc'.
> java.lang.Exception: Disallowed class(es) in the public API:
  - DefaultServiceRequestContext -> com.linecorp.armeria.internal.common.KeepAliveHandler
  - ServiceRequestContextBuilder -> com.linecorp.armeria.internal.common.KeepAliveHandler
```

Reply:
> the grace period as described above

I wonder that do you really need a grace period? As you know a `ServiceRequestContext` works at the request level. It seems like overkill to sophisticatedly control connection. I think it is enough to immediately send a GOAWAY because a GOAWAY with lastStreamId is already graceful. What do you think?

> DefaultServiceRequestContext methods can be called concurrently from multiple threads 

You can use the following pattern to avoid multi-threading issue:
```
if (eventLoop.inEventLoop()) {
   keepAlive.xxxx();
} else {
   eventLoop.execute(() -> keepAlive.xxxx());
}
```

> CI checks are unhappy:

You might link to `{@link KeepAliveHandler}` in the Javadoc. How about using `{@code KeepAliveHandler}`?

Reply:
Had a chat with @trustin, and I understood why a grace period is important. 
We also reached to an idea that adds the grace period option to `ServerBuilder` instead of `ServiceRequestContext.initiateConnectionShutdown()`.

- If a grace period configured by a `ServerBuilder` or `Flags` is positive
  - The first GOAWAY frame will be sent with stream ID 2^31-1
  - The second GOAWAY frame will be sent with the last stream ID after the grace period.
- If a grace period configured by a `ServerBuilder` or `Flags` is zero
  - The one GOAWAY frame will be sent with stream ID the last stream ID/

As the configuration is configured at a server level, this policy has to apply to all GOAWAY frames that are written by the Armeria server.

Reply:
@alexc-db, @ikhoon and I continued the discussion on this at [a Slack thread](https://line-armeria.slack.com/archives/C1NGPBUH2/p1627017213236500).

Reply:
@trustin @ikhoon I updated the change using ChannelPipeline::fireUserEventTriggered API. I diverted from the idea of writing GOAWAY from KeepAliveHandler - it seemed to be better done from the respective channel handlers (e.g. Http2ServerConnectionHandler::goAway does the job for HTTP/2), KeepAliveHandler stays focused on implementing ping/connection idle/connection age logic.

Please take a look and let me know if this is moving in the right direction. I'll follow up on the server level configuration if this looks good.

Reply:
Added server-level configuration and made a couple of extra changes to allow executing graceful connection shutdown logic from the close() of the HTTP/2 handler, so it's reused for all channel.close() calls on the server. 

However it's not easy to do the same with HTTP/1, so for now it's executed from userEventTriggered.

Note: according to [this SO answer](https://stackoverflow.com/a/46513136/16352922) once server has returned header with "Connection: close" it should reply 408 for all incoming requests so that clients could safely retry them. If that's the case - there's no grace period for in flight requests similar to [what's described in the HTTP/2 spec](https://datatracker.ietf.org/doc/html/rfc7540#section-6.8) and what the gracePeriod configuration added here is all about. I'm not sure if there's a better alternative - https://datatracker.ietf.org/doc/html/rfc2616#section-14.10 says:

> HTTP/1.1 defines the "close" connection option for the sender to signal that the connection will be closed after completion of the response.

So e.g. accepting in flight requests and returning responses with "Connection: close" multiple times would be inconsistent with the spec, if I'm reading it correctly. Current implementation of the grace period simply delays destruction of the keep alive handler (which triggers "Connection: close" header), maybe we should keep it this way.

Reply:
A possible option for HTTP/1: on grace period start: destroy keep alive handler (so that next response has "Connection: close" and no new requests are accepted), on grace period end: close connection immediately. A potential issue with it - graceful shutdown behavior will be different for HTTP/1 and HTTP/2.

HTTP/2 goes through the following states:
1. Grace period start: send GOAWAY 2^31-1, continue accepting new streams. client eventually receives GOAWAY but expects in flight requests to be processed.
2. Grace period end: send GOAWAY lastStreamId, new streams from client are rejected, wait for responses to be sent for already accepted streams.
3. Once all streams are closed - close the connection.

HTTP/1 would go through the following states:
1. Grace period start: send "Connection: close" w/ next headers, stop accepting incoming requests, wait for the current response to finish.
2. Grace period end: close the connection without waiting for any additional data to flush.

As you can see HTTP/1 state 1. and 2. would map to HTTP/2 state 2. and 3., which may be confusing. 

Current implementation:

HTTP/1
1. Grace period start: no-op
2. Grace period end: send "Connection: close" w/ next headers, stop accepting incoming requests, wait for the current response to finish.
3. Once response processing is finished connection will be closed

It maps better to current HTTP/2 states. 1. still accepts incoming requests, but doesn't signal clients anything (because there's no way it seems), so having grace period for HTTP/1 is kinda pointless. Let me know if you think there's a 3rd option.

Reply:
> HTTP/1 would go through the following states:
>
> 1. Grace period start: send "Connection: close" w/ next headers, stop accepting incoming requests, wait for the current response to finish.
> 2. Grace period end: close the connection without waiting for any additional data to flush.

If `ServiceRequestContext.initiateCloseShutdown()` guarantees to complete the current request and rejectcs new requests,
I don't think we can choose these states for HTTP/1.

> HTTP/1 would go through the following states:

> 1. Grace period start: send "Connection: close" w/ next headers, stop accepting incoming requests, wait for the current response to finish.
> 2. Grace period end: no-op.
> 3. Once response processing is finished connection will be closed

A grace period would be useful to avoid aborting in-flight requests. If a client uses pipelining[https://en.wikipedia.org/wiki/HTTP_pipelining], 
don't we need to accept the incoming requests during the grace period?

I propose the following states for HTTP/1 for leveraging grace periods.

> 1. Grace period start: send "Connection: close" w/ next headers, accept incoming requests that might be sent before receiving "Connection: close".
> 2. Grace period end: stop accepting incoming requests.
> 3. Once response processing is finished connection will be closed

Reply:
> If ServiceRequestContext.initiateCloseShutdown() guarantees to complete the current request and rejectcs new requests, I don't think we can choose these states for HTTP/1.

+1, I don't think it's a good option.

> A grace period would be useful to avoid aborting in-flight requests. If a client uses pipelining,
don't we need to accept the in-flight requests during the grace period?
>
> I propose the following states for HTTP/1 for leveraging grace periods.
>
> Grace period start: send "Connection: close" w/ next headers, accept incoming requests that might be sent before receiving "Connection: close".
> Grace period end: stop accepting incoming requests.
> Once response processing is finished connection will be closed

If we accept incoming requests that might be sent before receiving "Connection: close" we'll need to respond to them with "Connection: close" again. I think that'd violate HTTP/1 spec https://httpwg.org/specs/rfc7230.html#persistent.tear-down:

"A server that sends a "close" connection option MUST initiate a close of the connection (see below) after it sends the response containing "close". The server MUST NOT process any further requests received on that connection."

Spec doesn't special-case that for pipelining, it assumes that pipelining is only used for idempotent requests that can be safely retried https://datatracker.ietf.org/doc/html/rfc7230#section-6.3.2:

"A server MAY process a sequence of pipelined requests in parallel if they all have safe methods (Section 4.2.1 of [RFC7231]), but it MUST send the corresponding responses in the same order that the requests were received."

"A client that pipelines requests SHOULD retry unanswered requests if the connection closes before it receives all of the corresponding responses."

So my reading of the two quotes above is that server should close connection after the last response with "Connection: close" without processing in flight requests, they should be retried by the client.

Reply:
> So my reading of the two quotes above is that server should close connection after the last response with "Connection: close" without processing in flight requests, they should be retried by the client.

+1 You are right. A client who uses HTTP/1 pipelining SHOULD NOT expect normal responses after receiving "Connection: close".

By the way, I realized we have a bug on HTTP/1 pipelining, thanks to your investigation. #3723

Reply:
Happy to help!

After thinking more about this I feel like we should probably go with option 2 https://github.com/line/armeria/pull/3715#discussion_r676248551. Instantly destroy keep alive handler, send last response with "Connection: close" and close connection, add a note to the method / flag / server config description that grace period is not supported by HTTP/1 handler, so it'll behave as if grace period was set to 0. If user calls initiateConnectionShutdown it makes little sense to artificially delay actual connection shutdown. HTTP/1 spec doesn't support grace period as defined above, so we remove it.

Reply:
That sounds nice to me.